##### Copyright 2019 The TensorFlow Authors.

In [1]:
#@title Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

In [2]:
#@title MIT License
#
# Copyright (c) 2017 François Chollet
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.

# 使用 Keras 和 Tensorflow Hub 对电影评论进行文本分类

<table class="tfo-notebook-buttons" align="left">
  <td>
    <a target="_blank" href="https://tensorflow.google.cn/tutorials/keras/text_classification_with_hub"><img src="https://tensorflow.google.cn/images/tf_logo_32px.png" />在 tensorFlow.google.cn 上查看</a>
  </td>
  <td>
    <a target="_blank" href="https://colab.research.google.com/github/tensorflow/docs-l10n/blob/master/site/zh-cn/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://tensorflow.google.cn/images/colab_logo_32px.png" />在 Google Colab 中运行</a>
  </td>
  <td>
    <a target="_blank" href="https://github.com/tensorflow/docs-l10n/blob/master/site/zh-cn/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://tensorflow.google.cn/images/GitHub-Mark-32px.png" />在 GitHub 上查看源代码</a>
  </td>
  <td>
    <a href="https://storage.googleapis.com/tensorflow_docs/docs-l10n/site/zh-cn/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://tensorflow.google.cn/images/download_logo_32px.png" />下载 notebook</a>
  </td>
</table>

Note: 我们的 TensorFlow 社区翻译了这些文档。因为社区翻译是尽力而为， 所以无法保证它们是最准确的，并且反映了最新的
[官方英文文档](https://tensorflow.google.cn/?hl=en)。如果您有改进此翻译的建议， 请提交 pull request 到
[tensorflow/docs](https://github.com/tensorflow/docs) GitHub 仓库。要志愿地撰写或者审核译文，请加入
[docs-zh-cn@tensorflow.org Google Group](https://groups.google.com/a/tensorflow.org/forum/#!forum/docs-zh-cn)。

此笔记本（notebook）使用评论文本将影评分为*积极（positive）*或*消极（nagetive）*两类。这是一个*二元（binary）*或者二分类问题，一种重要且应用广泛的机器学习问题。

本教程演示了使用 Tensorflow Hub 和 Keras 进行迁移学习的基本应用。

我们将使用来源于[网络电影数据库（Internet Movie Database）](https://www.imdb.com/)的 [IMDB 数据集（IMDB dataset）](https://tensorflow.google.cn/api_docs/python/tf/keras/datasets/imdb)，其包含 50,000 条影评文本。从该数据集切割出的 25,000 条评论用作训练，另外 25,000 条用作测试。训练集与测试集是*平衡的（balanced）*，意味着它们包含相等数量的积极和消极评论。

此笔记本（notebook）使用了 [tf.keras](https://tensorflow.google.cn/guide/keras)，它是一个 Tensorflow 中用于构建和训练模型的高级API，此外还使用了 [TensorFlow Hub](https://tensorflow.google.cn/hub)，一个用于迁移学习的库和平台。有关使用 `tf.keras` 进行文本分类的更高级教程，请参阅 [MLCC文本分类指南（MLCC Text Classification Guide）](https://developers.google.com/machine-learning/guides/text-classification/)。

In [3]:
import numpy as np

import tensorflow as tf

!pip install -q tensorflow-hub
!pip install -q tfds-nightly
import tensorflow_hub as hub
import tensorflow_datasets as tfds

print("Version: ", tf.__version__)
print("Eager mode: ", tf.executing_eagerly())
print("Hub version: ", hub.__version__)
print("GPU is", "available" if tf.config.experimental.list_physical_devices("GPU") else "NOT AVAILABLE")

Version:  2.3.0
Eager mode:  True
Hub version:  0.8.0
GPU is available


## 下载 IMDB 数据集
IMDB数据集可以在 [Tensorflow 数据集](https://github.com/tensorflow/datasets)处获取。以下代码将 IMDB 数据集下载至您的机器（或 colab 运行时环境）中：

In [4]:
# 将训练集分割成 60% 和 40%，从而最终我们将得到 15,000 个训练样本
# 10,000 个验证样本以及 25,000 个测试样本。
train_data, validation_data, test_data = tfds.load(
    name="imdb_reviews", 
    split=('train[:60%]', 'train[60%:]', 'test'),
    as_supervised=True)

Shuffling and writing examples to /home/kbuilder/tensorflow_datasets/imdb_reviews/plain_text/1.0.0.incompleteO2CUAB/imdb_reviews-train.tfrecord


Shuffling and writing examples to /home/kbuilder/tensorflow_datasets/imdb_reviews/plain_text/1.0.0.incompleteO2CUAB/imdb_reviews-test.tfrecord


Shuffling and writing examples to /home/kbuilder/tensorflow_datasets/imdb_reviews/plain_text/1.0.0.incompleteO2CUAB/imdb_reviews-unsupervised.tfrecord
Dataset imdb_reviews downloaded and prepared to /home/kbuilder/tensorflow_datasets/imdb_reviews/plain_text/1.0.0. Subsequent calls will reuse this data.


## 探索数据

让我们花一点时间来了解数据的格式。每一个样本都是一个表示电影评论和相应标签的句子。该句子不以任何方式进行预处理。标签是一个值为 0 或 1 的整数，其中 0 代表消极评论，1 代表积极评论。

我们来打印下前十个样本。

In [5]:
train_examples_batch, train_labels_batch = next(iter(train_data.batch(10)))
train_examples_batch

<tf.Tensor: shape=(10,), dtype=string, numpy=
array([b"This was an absolutely terrible movie. Don't be lured in by Christopher Walken or Michael Ironside. Both are great actors, but this must simply be their worst role in history. Even their great acting could not redeem this movie's ridiculous storyline. This movie is an early nineties US propaganda piece. The most pathetic scenes were those when the Columbian rebels were making their cases for revolutions. Maria Conchita Alonso appeared phony, and her pseudo-love affair with Walken was nothing but a pathetic emotional plug in a movie that was devoid of any real meaning. I am disappointed that there are movies like this, ruining actor's like Christopher Walken's good name. I could barely sit through it.",
       b'I have been known to fall asleep during films, but this is usually due to a combination of things including, really tired, being warm and comfortable on the sette and having just eaten a lot. However on this occasion I fell 

我们再打印下前十个标签。

In [6]:
train_labels_batch

<tf.Tensor: shape=(10,), dtype=int64, numpy=array([0, 0, 0, 1, 1, 1, 0, 0, 0, 0])>

## 构建模型

神经网络由堆叠的层来构建，这需要从三个主要方面来进行体系结构决策：

* 如何表示文本？
* 模型里有多少层？
* 每个层里有多少*隐层单元（hidden units）*？

本示例中，输入数据由句子组成。预测的标签为 0 或 1。

表示文本的一种方式是将句子转换为嵌入向量（embeddings vectors）。我们可以使用一个预先训练好的文本嵌入（text embedding）作为首层，这将具有三个优点：

* 我们不必担心文本预处理
* 我们可以从迁移学习中受益
* 嵌入具有固定长度，更易于处理

针对此示例我们将使用 [TensorFlow Hub](https://tensorflow.google.cn/hub) 中名为 [google/tf2-preview/gnews-swivel-20dim/1](https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1) 的一种**预训练文本嵌入（text embedding）模型** 。

为了达到本教程的目的还有其他三种预训练模型可供测试：

* [google/tf2-preview/gnews-swivel-20dim-with-oov/1](https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim-with-oov/1) ——类似 [google/tf2-preview/gnews-swivel-20dim/1](https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1)，但 2.5%的词汇转换为未登录词桶（OOV buckets）。如果任务的词汇与模型的词汇没有完全重叠，这将会有所帮助。
* [google/tf2-preview/nnlm-en-dim50/1](https://tfhub.dev/google/tf2-preview/nnlm-en-dim50/1) ——一个拥有约 1M 词汇量且维度为 50 的更大的模型。
* [google/tf2-preview/nnlm-en-dim128/1](https://tfhub.dev/google/tf2-preview/nnlm-en-dim128/1) ——拥有约 1M 词汇量且维度为128的更大的模型。

让我们首先创建一个使用 Tensorflow Hub 模型嵌入（embed）语句的Keras层，并在几个输入样本中进行尝试。请注意无论输入文本的长度如何，嵌入（embeddings）输出的形状都是：`(num_examples, embedding_dimension)`。


In [7]:
embedding = "https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1"
hub_layer = hub.KerasLayer(embedding, input_shape=[], 
                           dtype=tf.string, trainable=True)
hub_layer(train_examples_batch[:3])

<tf.Tensor: shape=(3, 20), dtype=float32, numpy=
array([[ 1.765786  , -3.882232  ,  3.9134233 , -1.5557289 , -3.3362343 ,
        -1.7357955 , -1.9954445 ,  1.2989551 ,  5.081598  , -1.1041286 ,
        -2.0503852 , -0.72675157, -0.65675956,  0.24436149, -3.7208383 ,
         2.0954835 ,  2.2969332 , -2.0689783 , -2.9489717 , -1.1315987 ],
       [ 1.8804485 , -2.5852382 ,  3.4066997 ,  1.0982676 , -4.056685  ,
        -4.891284  , -2.785554  ,  1.3874227 ,  3.8476458 , -0.9256538 ,
        -1.896706  ,  1.2113281 ,  0.11474707,  0.76209456, -4.8791065 ,
         2.906149  ,  4.7087674 , -2.3652055 , -3.5015898 , -1.6390051 ],
       [ 0.71152234, -0.6353217 ,  1.7385626 , -1.1168286 , -0.5451594 ,
        -1.1808156 ,  0.09504455,  1.4653089 ,  0.66059524,  0.79308075,
        -2.2268345 ,  0.07446612, -1.4075904 , -0.70645386, -1.907037  ,
         1.4419787 ,  1.9551861 , -0.42660055, -2.8022065 ,  0.43727064]],
      dtype=float32)>

现在让我们构建完整模型：

In [8]:
model = tf.keras.Sequential()
model.add(hub_layer)
model.add(tf.keras.layers.Dense(16, activation='relu'))
model.add(tf.keras.layers.Dense(1))

model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
keras_layer (KerasLayer)     (None, 20)                400020    
_________________________________________________________________
dense (Dense)                (None, 16)                336       
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 17        
Total params: 400,373
Trainable params: 400,373
Non-trainable params: 0
_________________________________________________________________


层按顺序堆叠以构建分类器：

1. 第一层是 Tensorflow Hub 层。这一层使用一个预训练的保存好的模型来将句子映射为嵌入向量（embedding vector）。我们所使用的预训练文本嵌入（embedding）模型([google/tf2-preview/gnews-swivel-20dim/1](https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1))将句子切割为符号，嵌入（embed）每个符号然后进行合并。最终得到的维度是：`(num_examples, embedding_dimension)`。
2. 该定长输出向量通过一个有 16 个隐层单元的全连接层（`Dense`）进行管道传输。
3. 最后一层与单个输出结点紧密相连。使用 `Sigmoid` 激活函数，其函数值为介于 0 与 1 之间的浮点数，表示概率或置信水平。

让我们编译模型。

### 损失函数与优化器

一个模型需要损失函数和优化器来进行训练。由于这是一个二分类问题且模型输出概率值（一个使用 sigmoid 激活函数的单一单元层），我们将使用 `binary_crossentropy` 损失函数。

这不是损失函数的唯一选择，例如，您可以选择 `mean_squared_error` 。但是，一般来说 `binary_crossentropy` 更适合处理概率——它能够度量概率分布之间的“距离”，或者在我们的示例中，指的是度量 ground-truth 分布与预测值之间的“距离”。

稍后，当我们研究回归问题（例如，预测房价）时，我们将介绍如何使用另一种叫做均方误差的损失函数。

现在，配置模型来使用优化器和损失函数：

In [9]:
model.compile(optimizer='adam',
              loss=tf.keras.losses.BinaryCrossentropy(from_logits=True),
              metrics=['accuracy'])

## 训练模型

以 512 个样本的 mini-batch 大小迭代 20 个 epoch 来训练模型。 这是指对 `x_train` 和 `y_train` 张量中所有样本的的 20 次迭代。在训练过程中，监测来自验证集的 10,000 个样本上的损失值（loss）和准确率（accuracy）：

In [10]:
history = model.fit(train_data.shuffle(10000).batch(512),
                    epochs=20,
                    validation_data=validation_data.batch(512),
                    verbose=1)

Epoch 1/20


 1/30 [>.............................] - ETA: 0s - loss: 2.5708 - accuracy: 0.4863

 3/30 [==>...........................] - ETA: 0s - loss: 2.3588 - accuracy: 0.5052

 5/30 [====>.........................] - ETA: 0s - loss: 2.2768 - accuracy: 0.5012

 7/30 [======>.......................] - ETA: 0s - loss: 2.2244 - accuracy: 0.4978

 9/30 [========>.....................] - ETA: 0s - loss: 2.1426 - accuracy: 0.4976

11/30 [==========>...................] - ETA: 0s - loss: 2.0604 - accuracy: 0.5009

13/30 [============>.................] - ETA: 0s - loss: 1.9952 - accuracy: 0.4992

15/30 [==============>...............] - ETA: 0s - loss: 1.9271 - accuracy: 0.4982

17/30 [================>.............] - ETA: 0s - loss: 1.8566 - accuracy: 0.4997

19/30 [==================>...........] - ETA: 0s - loss: 1.7827 - accuracy: 0.5025

21/30 [====================>.........] - ETA: 0s - loss: 1.7169 - accuracy: 0.5032

23/30 [======================>.......] - ETA: 0s - loss: 1.6580 - accuracy: 0.5019

25/30 [========================>.....] - ETA: 0s - loss: 1.6009 - accuracy: 0.5016

27/30 [==========================>...] - ETA: 0s - loss: 1.5449 - accuracy: 0.5021

29/30 [============================>.] - ETA: 0s - loss: 1.4918 - accuracy: 0.5038

30/30 [==============================] - 2s 56ms/step - loss: 1.4843 - accuracy: 0.5042 - val_loss: 0.7258 - val_accuracy: 0.5346


Epoch 2/20


 1/30 [>.............................] - ETA: 0s - loss: 0.7366 - accuracy: 0.5176

 3/30 [==>...........................] - ETA: 0s - loss: 0.7150 - accuracy: 0.5501

 5/30 [====>.........................] - ETA: 0s - loss: 0.7065 - accuracy: 0.5570

 7/30 [======>.......................] - ETA: 0s - loss: 0.6920 - accuracy: 0.5737

 9/30 [========>.....................] - ETA: 0s - loss: 0.6894 - accuracy: 0.5825

11/30 [==========>...................] - ETA: 0s - loss: 0.6905 - accuracy: 0.5868

13/30 [============>.................] - ETA: 0s - loss: 0.6857 - accuracy: 0.5944

15/30 [==============>...............] - ETA: 0s - loss: 0.6837 - accuracy: 0.5990

17/30 [================>.............] - ETA: 0s - loss: 0.6837 - accuracy: 0.6009

19/30 [==================>...........] - ETA: 0s - loss: 0.6833 - accuracy: 0.6035

21/30 [====================>.........] - ETA: 0s - loss: 0.6816 - accuracy: 0.6062

23/30 [======================>.......] - ETA: 0s - loss: 0.6810 - accuracy: 0.6080

25/30 [========================>.....] - ETA: 0s - loss: 0.6805 - accuracy: 0.6098

27/30 [==========================>...] - ETA: 0s - loss: 0.6782 - accuracy: 0.6127

29/30 [============================>.] - ETA: 0s - loss: 0.6763 - accuracy: 0.6142

30/30 [==============================] - 2s 51ms/step - loss: 0.6759 - accuracy: 0.6145 - val_loss: 0.6529 - val_accuracy: 0.6318


Epoch 3/20


 1/30 [>.............................] - ETA: 0s - loss: 0.6532 - accuracy: 0.5996

 3/30 [==>...........................] - ETA: 0s - loss: 0.6448 - accuracy: 0.6315

 5/30 [====>.........................] - ETA: 0s - loss: 0.6543 - accuracy: 0.6168

 7/30 [======>.......................] - ETA: 0s - loss: 0.6554 - accuracy: 0.6144

 9/30 [========>.....................] - ETA: 0s - loss: 0.6493 - accuracy: 0.6215

11/30 [==========>...................] - ETA: 0s - loss: 0.6452 - accuracy: 0.6284

13/30 [============>.................] - ETA: 0s - loss: 0.6407 - accuracy: 0.6349

15/30 [==============>...............] - ETA: 0s - loss: 0.6410 - accuracy: 0.6342

17/30 [================>.............] - ETA: 0s - loss: 0.6395 - accuracy: 0.6340

19/30 [==================>...........] - ETA: 0s - loss: 0.6391 - accuracy: 0.6326

21/30 [====================>.........] - ETA: 0s - loss: 0.6385 - accuracy: 0.6335

23/30 [======================>.......] - ETA: 0s - loss: 0.6377 - accuracy: 0.6345

25/30 [========================>.....] - ETA: 0s - loss: 0.6375 - accuracy: 0.6341

27/30 [==========================>...] - ETA: 0s - loss: 0.6365 - accuracy: 0.6336

29/30 [============================>.] - ETA: 0s - loss: 0.6365 - accuracy: 0.6329

30/30 [==============================] - 2s 50ms/step - loss: 0.6361 - accuracy: 0.6339 - val_loss: 0.6268 - val_accuracy: 0.6508


Epoch 4/20


 1/30 [>.............................] - ETA: 0s - loss: 0.6190 - accuracy: 0.6504

 3/30 [==>...........................] - ETA: 0s - loss: 0.6238 - accuracy: 0.6530

 5/30 [====>.........................] - ETA: 0s - loss: 0.6278 - accuracy: 0.6500

 7/30 [======>.......................] - ETA: 0s - loss: 0.6264 - accuracy: 0.6501

 9/30 [========>.....................] - ETA: 0s - loss: 0.6267 - accuracy: 0.6461

11/30 [==========>...................] - ETA: 0s - loss: 0.6246 - accuracy: 0.6436

13/30 [============>.................] - ETA: 0s - loss: 0.6203 - accuracy: 0.6481

15/30 [==============>...............] - ETA: 0s - loss: 0.6180 - accuracy: 0.6518

17/30 [================>.............] - ETA: 0s - loss: 0.6150 - accuracy: 0.6556

19/30 [==================>...........] - ETA: 0s - loss: 0.6128 - accuracy: 0.6592

21/30 [====================>.........] - ETA: 0s - loss: 0.6107 - accuracy: 0.6614

23/30 [======================>.......] - ETA: 0s - loss: 0.6114 - accuracy: 0.6619

25/30 [========================>.....] - ETA: 0s - loss: 0.6107 - accuracy: 0.6627

27/30 [==========================>...] - ETA: 0s - loss: 0.6085 - accuracy: 0.6655

29/30 [============================>.] - ETA: 0s - loss: 0.6088 - accuracy: 0.6656

30/30 [==============================] - 1s 49ms/step - loss: 0.6086 - accuracy: 0.6663 - val_loss: 0.6010 - val_accuracy: 0.6707


Epoch 5/20


 1/30 [>.............................] - ETA: 0s - loss: 0.6193 - accuracy: 0.6582

 3/30 [==>...........................] - ETA: 0s - loss: 0.6042 - accuracy: 0.6660

 5/30 [====>.........................] - ETA: 0s - loss: 0.5930 - accuracy: 0.6695

 7/30 [======>.......................] - ETA: 0s - loss: 0.5922 - accuracy: 0.6769

 9/30 [========>.....................] - ETA: 0s - loss: 0.5870 - accuracy: 0.6845

11/30 [==========>...................] - ETA: 0s - loss: 0.5860 - accuracy: 0.6829

13/30 [============>.................] - ETA: 0s - loss: 0.5826 - accuracy: 0.6858

15/30 [==============>...............] - ETA: 0s - loss: 0.5804 - accuracy: 0.6884

17/30 [================>.............] - ETA: 0s - loss: 0.5816 - accuracy: 0.6877

19/30 [==================>...........] - ETA: 0s - loss: 0.5819 - accuracy: 0.6871

21/30 [====================>.........] - ETA: 0s - loss: 0.5815 - accuracy: 0.6861

23/30 [======================>.......] - ETA: 0s - loss: 0.5808 - accuracy: 0.6865

25/30 [========================>.....] - ETA: 0s - loss: 0.5809 - accuracy: 0.6867

27/30 [==========================>...] - ETA: 0s - loss: 0.5819 - accuracy: 0.6854

29/30 [============================>.] - ETA: 0s - loss: 0.5804 - accuracy: 0.6870

30/30 [==============================] - 1s 49ms/step - loss: 0.5801 - accuracy: 0.6875 - val_loss: 0.5737 - val_accuracy: 0.6941


Epoch 6/20


 1/30 [>.............................] - ETA: 0s - loss: 0.5472 - accuracy: 0.6992

 3/30 [==>...........................] - ETA: 0s - loss: 0.5538 - accuracy: 0.7103

 5/30 [====>.........................] - ETA: 0s - loss: 0.5556 - accuracy: 0.7109

 7/30 [======>.......................] - ETA: 0s - loss: 0.5590 - accuracy: 0.7081

 9/30 [========>.....................] - ETA: 0s - loss: 0.5612 - accuracy: 0.7086

11/30 [==========>...................] - ETA: 0s - loss: 0.5594 - accuracy: 0.7093

13/30 [============>.................] - ETA: 0s - loss: 0.5570 - accuracy: 0.7105

15/30 [==============>...............] - ETA: 0s - loss: 0.5562 - accuracy: 0.7095

17/30 [================>.............] - ETA: 0s - loss: 0.5541 - accuracy: 0.7137

19/30 [==================>...........] - ETA: 0s - loss: 0.5536 - accuracy: 0.7145

21/30 [====================>.........] - ETA: 0s - loss: 0.5518 - accuracy: 0.7140

23/30 [======================>.......] - ETA: 0s - loss: 0.5512 - accuracy: 0.7131

25/30 [========================>.....] - ETA: 0s - loss: 0.5494 - accuracy: 0.7143

27/30 [==========================>...] - ETA: 0s - loss: 0.5484 - accuracy: 0.7141

29/30 [============================>.] - ETA: 0s - loss: 0.5486 - accuracy: 0.7144

30/30 [==============================] - 1s 48ms/step - loss: 0.5484 - accuracy: 0.7145 - val_loss: 0.5437 - val_accuracy: 0.7173


Epoch 7/20


 1/30 [>.............................] - ETA: 0s - loss: 0.5608 - accuracy: 0.7109

 3/30 [==>...........................] - ETA: 0s - loss: 0.5366 - accuracy: 0.7279

 5/30 [====>.........................] - ETA: 0s - loss: 0.5340 - accuracy: 0.7262

 7/30 [======>.......................] - ETA: 0s - loss: 0.5309 - accuracy: 0.7355

 9/30 [========>.....................] - ETA: 0s - loss: 0.5288 - accuracy: 0.7359

11/30 [==========>...................] - ETA: 0s - loss: 0.5296 - accuracy: 0.7321

13/30 [============>.................] - ETA: 0s - loss: 0.5279 - accuracy: 0.7374

15/30 [==============>...............] - ETA: 0s - loss: 0.5248 - accuracy: 0.7363

17/30 [================>.............] - ETA: 0s - loss: 0.5226 - accuracy: 0.7372

19/30 [==================>...........] - ETA: 0s - loss: 0.5199 - accuracy: 0.7398

21/30 [====================>.........] - ETA: 0s - loss: 0.5188 - accuracy: 0.7403

23/30 [======================>.......] - ETA: 0s - loss: 0.5172 - accuracy: 0.7407

25/30 [========================>.....] - ETA: 0s - loss: 0.5155 - accuracy: 0.7400

27/30 [==========================>...] - ETA: 0s - loss: 0.5137 - accuracy: 0.7421

29/30 [============================>.] - ETA: 0s - loss: 0.5139 - accuracy: 0.7419

30/30 [==============================] - 1s 50ms/step - loss: 0.5136 - accuracy: 0.7419 - val_loss: 0.5110 - val_accuracy: 0.7460


Epoch 8/20


 1/30 [>.............................] - ETA: 0s - loss: 0.4901 - accuracy: 0.7617

 3/30 [==>...........................] - ETA: 0s - loss: 0.4959 - accuracy: 0.7526

 5/30 [====>.........................] - ETA: 0s - loss: 0.4990 - accuracy: 0.7520

 7/30 [======>.......................] - ETA: 0s - loss: 0.4962 - accuracy: 0.7573

 9/30 [========>.....................] - ETA: 0s - loss: 0.4896 - accuracy: 0.7617

11/30 [==========>...................] - ETA: 0s - loss: 0.4916 - accuracy: 0.7589

13/30 [============>.................] - ETA: 0s - loss: 0.4866 - accuracy: 0.7634

15/30 [==============>...............] - ETA: 0s - loss: 0.4827 - accuracy: 0.7673

17/30 [================>.............] - ETA: 0s - loss: 0.4819 - accuracy: 0.7677

19/30 [==================>...........] - ETA: 0s - loss: 0.4815 - accuracy: 0.7679

21/30 [====================>.........] - ETA: 0s - loss: 0.4792 - accuracy: 0.7706

23/30 [======================>.......] - ETA: 0s - loss: 0.4797 - accuracy: 0.7694

25/30 [========================>.....] - ETA: 0s - loss: 0.4787 - accuracy: 0.7700

27/30 [==========================>...] - ETA: 0s - loss: 0.4777 - accuracy: 0.7689

29/30 [============================>.] - ETA: 0s - loss: 0.4761 - accuracy: 0.7697

30/30 [==============================] - 1s 48ms/step - loss: 0.4759 - accuracy: 0.7701 - val_loss: 0.4773 - val_accuracy: 0.7636


Epoch 9/20


 1/30 [>.............................] - ETA: 0s - loss: 0.4506 - accuracy: 0.8066

 3/30 [==>...........................] - ETA: 0s - loss: 0.4503 - accuracy: 0.7839

 5/30 [====>.........................] - ETA: 0s - loss: 0.4542 - accuracy: 0.7844

 7/30 [======>.......................] - ETA: 0s - loss: 0.4591 - accuracy: 0.7793

 9/30 [========>.....................] - ETA: 0s - loss: 0.4560 - accuracy: 0.7828

11/30 [==========>...................] - ETA: 0s - loss: 0.4546 - accuracy: 0.7869

13/30 [============>.................] - ETA: 0s - loss: 0.4535 - accuracy: 0.7880

15/30 [==============>...............] - ETA: 0s - loss: 0.4508 - accuracy: 0.7882

17/30 [================>.............] - ETA: 0s - loss: 0.4476 - accuracy: 0.7901

19/30 [==================>...........] - ETA: 0s - loss: 0.4451 - accuracy: 0.7924

21/30 [====================>.........] - ETA: 0s - loss: 0.4430 - accuracy: 0.7935

23/30 [======================>.......] - ETA: 0s - loss: 0.4401 - accuracy: 0.7945

25/30 [========================>.....] - ETA: 0s - loss: 0.4390 - accuracy: 0.7962

27/30 [==========================>...] - ETA: 0s - loss: 0.4367 - accuracy: 0.7979

29/30 [============================>.] - ETA: 0s - loss: 0.4371 - accuracy: 0.7975

30/30 [==============================] - 1s 49ms/step - loss: 0.4365 - accuracy: 0.7982 - val_loss: 0.4414 - val_accuracy: 0.7862


Epoch 10/20


 1/30 [>.............................] - ETA: 0s - loss: 0.4234 - accuracy: 0.8047

 3/30 [==>...........................] - ETA: 0s - loss: 0.3995 - accuracy: 0.8210

 5/30 [====>.........................] - ETA: 0s - loss: 0.3989 - accuracy: 0.8191

 7/30 [======>.......................] - ETA: 0s - loss: 0.4006 - accuracy: 0.8145

 9/30 [========>.....................] - ETA: 0s - loss: 0.4007 - accuracy: 0.8173

11/30 [==========>...................] - ETA: 0s - loss: 0.4013 - accuracy: 0.8184

13/30 [============>.................] - ETA: 0s - loss: 0.4042 - accuracy: 0.8158

15/30 [==============>...............] - ETA: 0s - loss: 0.4008 - accuracy: 0.8178

17/30 [================>.............] - ETA: 0s - loss: 0.4007 - accuracy: 0.8165

19/30 [==================>...........] - ETA: 0s - loss: 0.4009 - accuracy: 0.8161

21/30 [====================>.........] - ETA: 0s - loss: 0.3996 - accuracy: 0.8169

23/30 [======================>.......] - ETA: 0s - loss: 0.3988 - accuracy: 0.8175

25/30 [========================>.....] - ETA: 0s - loss: 0.3977 - accuracy: 0.8184

27/30 [==========================>...] - ETA: 0s - loss: 0.3962 - accuracy: 0.8197

29/30 [============================>.] - ETA: 0s - loss: 0.3948 - accuracy: 0.8204

30/30 [==============================] - 1s 49ms/step - loss: 0.3947 - accuracy: 0.8204 - val_loss: 0.4078 - val_accuracy: 0.8112


Epoch 11/20


 1/30 [>.............................] - ETA: 0s - loss: 0.3466 - accuracy: 0.8535

 3/30 [==>...........................] - ETA: 0s - loss: 0.3635 - accuracy: 0.8398

 5/30 [====>.........................] - ETA: 0s - loss: 0.3683 - accuracy: 0.8410

 7/30 [======>.......................] - ETA: 0s - loss: 0.3688 - accuracy: 0.8407

 9/30 [========>.....................] - ETA: 0s - loss: 0.3676 - accuracy: 0.8401

11/30 [==========>...................] - ETA: 0s - loss: 0.3656 - accuracy: 0.8375

13/30 [============>.................] - ETA: 0s - loss: 0.3635 - accuracy: 0.8380

15/30 [==============>...............] - ETA: 0s - loss: 0.3610 - accuracy: 0.8400

17/30 [================>.............] - ETA: 0s - loss: 0.3590 - accuracy: 0.8404

19/30 [==================>...........] - ETA: 0s - loss: 0.3577 - accuracy: 0.8409

21/30 [====================>.........] - ETA: 0s - loss: 0.3564 - accuracy: 0.8417

23/30 [======================>.......] - ETA: 0s - loss: 0.3553 - accuracy: 0.8427

25/30 [========================>.....] - ETA: 0s - loss: 0.3551 - accuracy: 0.8424

27/30 [==========================>...] - ETA: 0s - loss: 0.3560 - accuracy: 0.8415

29/30 [============================>.] - ETA: 0s - loss: 0.3560 - accuracy: 0.8417

30/30 [==============================] - 1s 48ms/step - loss: 0.3555 - accuracy: 0.8421 - val_loss: 0.3808 - val_accuracy: 0.8225


Epoch 12/20


 1/30 [>.............................] - ETA: 0s - loss: 0.3394 - accuracy: 0.8418

 3/30 [==>...........................] - ETA: 0s - loss: 0.3345 - accuracy: 0.8522

 5/30 [====>.........................] - ETA: 0s - loss: 0.3329 - accuracy: 0.8523

 7/30 [======>.......................] - ETA: 0s - loss: 0.3324 - accuracy: 0.8546

 9/30 [========>.....................] - ETA: 0s - loss: 0.3308 - accuracy: 0.8524

11/30 [==========>...................] - ETA: 0s - loss: 0.3301 - accuracy: 0.8548

13/30 [============>.................] - ETA: 0s - loss: 0.3287 - accuracy: 0.8559

15/30 [==============>...............] - ETA: 0s - loss: 0.3294 - accuracy: 0.8569

17/30 [================>.............] - ETA: 0s - loss: 0.3300 - accuracy: 0.8555

19/30 [==================>...........] - ETA: 0s - loss: 0.3303 - accuracy: 0.8559

21/30 [====================>.........] - ETA: 0s - loss: 0.3280 - accuracy: 0.8576

23/30 [======================>.......] - ETA: 0s - loss: 0.3264 - accuracy: 0.8579

25/30 [========================>.....] - ETA: 0s - loss: 0.3252 - accuracy: 0.8593

27/30 [==========================>...] - ETA: 0s - loss: 0.3256 - accuracy: 0.8587

29/30 [============================>.] - ETA: 0s - loss: 0.3239 - accuracy: 0.8604

30/30 [==============================] - 1s 50ms/step - loss: 0.3235 - accuracy: 0.8605 - val_loss: 0.3604 - val_accuracy: 0.8333


Epoch 13/20


 1/30 [>.............................] - ETA: 0s - loss: 0.3112 - accuracy: 0.8633

 3/30 [==>...........................] - ETA: 0s - loss: 0.3122 - accuracy: 0.8574

 5/30 [====>.........................] - ETA: 0s - loss: 0.3064 - accuracy: 0.8645

 7/30 [======>.......................] - ETA: 0s - loss: 0.3073 - accuracy: 0.8658

 9/30 [========>.....................] - ETA: 0s - loss: 0.3054 - accuracy: 0.8694

11/30 [==========>...................] - ETA: 0s - loss: 0.3035 - accuracy: 0.8699

13/30 [============>.................] - ETA: 0s - loss: 0.3008 - accuracy: 0.8723

15/30 [==============>...............] - ETA: 0s - loss: 0.2978 - accuracy: 0.8738

17/30 [================>.............] - ETA: 0s - loss: 0.2966 - accuracy: 0.8742

19/30 [==================>...........] - ETA: 0s - loss: 0.2960 - accuracy: 0.8735

21/30 [====================>.........] - ETA: 0s - loss: 0.2969 - accuracy: 0.8741

23/30 [======================>.......] - ETA: 0s - loss: 0.2954 - accuracy: 0.8748

25/30 [========================>.....] - ETA: 0s - loss: 0.2946 - accuracy: 0.8764

27/30 [==========================>...] - ETA: 0s - loss: 0.2959 - accuracy: 0.8749

29/30 [============================>.] - ETA: 0s - loss: 0.2964 - accuracy: 0.8736

30/30 [==============================] - 1s 50ms/step - loss: 0.2963 - accuracy: 0.8738 - val_loss: 0.3446 - val_accuracy: 0.8441


Epoch 14/20


 1/30 [>.............................] - ETA: 0s - loss: 0.2885 - accuracy: 0.8633

 3/30 [==>...........................] - ETA: 0s - loss: 0.3024 - accuracy: 0.8633

 5/30 [====>.........................] - ETA: 0s - loss: 0.2980 - accuracy: 0.8676

 7/30 [======>.......................] - ETA: 0s - loss: 0.2954 - accuracy: 0.8733

 9/30 [========>.....................] - ETA: 0s - loss: 0.2910 - accuracy: 0.8761

11/30 [==========>...................] - ETA: 0s - loss: 0.2860 - accuracy: 0.8794

13/30 [============>.................] - ETA: 0s - loss: 0.2878 - accuracy: 0.8777

15/30 [==============>...............] - ETA: 0s - loss: 0.2830 - accuracy: 0.8796

17/30 [================>.............] - ETA: 0s - loss: 0.2820 - accuracy: 0.8802

19/30 [==================>...........] - ETA: 0s - loss: 0.2804 - accuracy: 0.8825

21/30 [====================>.........] - ETA: 0s - loss: 0.2778 - accuracy: 0.8848

23/30 [======================>.......] - ETA: 0s - loss: 0.2761 - accuracy: 0.8847

25/30 [========================>.....] - ETA: 0s - loss: 0.2753 - accuracy: 0.8852

27/30 [==========================>...] - ETA: 0s - loss: 0.2740 - accuracy: 0.8859

29/30 [============================>.] - ETA: 0s - loss: 0.2729 - accuracy: 0.8863

30/30 [==============================] - 1s 49ms/step - loss: 0.2731 - accuracy: 0.8861 - val_loss: 0.3324 - val_accuracy: 0.8545


Epoch 15/20


 1/30 [>.............................] - ETA: 0s - loss: 0.2691 - accuracy: 0.8945

 3/30 [==>...........................] - ETA: 0s - loss: 0.2465 - accuracy: 0.9089

 5/30 [====>.........................] - ETA: 0s - loss: 0.2549 - accuracy: 0.8992

 7/30 [======>.......................] - ETA: 0s - loss: 0.2600 - accuracy: 0.8965

 9/30 [========>.....................] - ETA: 0s - loss: 0.2558 - accuracy: 0.8974

11/30 [==========>...................] - ETA: 0s - loss: 0.2555 - accuracy: 0.8967

13/30 [============>.................] - ETA: 0s - loss: 0.2566 - accuracy: 0.8972

15/30 [==============>...............] - ETA: 0s - loss: 0.2542 - accuracy: 0.8975

17/30 [================>.............] - ETA: 0s - loss: 0.2533 - accuracy: 0.8973

19/30 [==================>...........] - ETA: 0s - loss: 0.2533 - accuracy: 0.8966

21/30 [====================>.........] - ETA: 0s - loss: 0.2546 - accuracy: 0.8958

23/30 [======================>.......] - ETA: 0s - loss: 0.2524 - accuracy: 0.8978

25/30 [========================>.....] - ETA: 0s - loss: 0.2526 - accuracy: 0.8981

27/30 [==========================>...] - ETA: 0s - loss: 0.2531 - accuracy: 0.8968

29/30 [============================>.] - ETA: 0s - loss: 0.2527 - accuracy: 0.8977

30/30 [==============================] - 1s 49ms/step - loss: 0.2532 - accuracy: 0.8971 - val_loss: 0.3228 - val_accuracy: 0.8585


Epoch 16/20


 1/30 [>.............................] - ETA: 0s - loss: 0.2321 - accuracy: 0.9102

 3/30 [==>...........................] - ETA: 0s - loss: 0.2336 - accuracy: 0.9108

 5/30 [====>.........................] - ETA: 0s - loss: 0.2353 - accuracy: 0.9078

 7/30 [======>.......................] - ETA: 0s - loss: 0.2448 - accuracy: 0.8987

 9/30 [========>.....................] - ETA: 0s - loss: 0.2453 - accuracy: 0.8995

11/30 [==========>...................] - ETA: 0s - loss: 0.2430 - accuracy: 0.9006

13/30 [============>.................] - ETA: 0s - loss: 0.2430 - accuracy: 0.9010

15/30 [==============>...............] - ETA: 0s - loss: 0.2426 - accuracy: 0.9034

17/30 [================>.............] - ETA: 0s - loss: 0.2416 - accuracy: 0.9037

19/30 [==================>...........] - ETA: 0s - loss: 0.2398 - accuracy: 0.9048

21/30 [====================>.........] - ETA: 0s - loss: 0.2396 - accuracy: 0.9048

23/30 [======================>.......] - ETA: 0s - loss: 0.2400 - accuracy: 0.9042

25/30 [========================>.....] - ETA: 0s - loss: 0.2386 - accuracy: 0.9045

27/30 [==========================>...] - ETA: 0s - loss: 0.2376 - accuracy: 0.9050

29/30 [============================>.] - ETA: 0s - loss: 0.2358 - accuracy: 0.9062

30/30 [==============================] - 1s 50ms/step - loss: 0.2359 - accuracy: 0.9061 - val_loss: 0.3155 - val_accuracy: 0.8601


Epoch 17/20


 1/30 [>.............................] - ETA: 0s - loss: 0.2409 - accuracy: 0.8945

 3/30 [==>...........................] - ETA: 0s - loss: 0.2221 - accuracy: 0.9062

 5/30 [====>.........................] - ETA: 0s - loss: 0.2203 - accuracy: 0.9102

 7/30 [======>.......................] - ETA: 0s - loss: 0.2203 - accuracy: 0.9116

 9/30 [========>.....................] - ETA: 0s - loss: 0.2217 - accuracy: 0.9132

11/30 [==========>...................] - ETA: 0s - loss: 0.2179 - accuracy: 0.9150

13/30 [============>.................] - ETA: 0s - loss: 0.2176 - accuracy: 0.9148

15/30 [==============>...............] - ETA: 0s - loss: 0.2190 - accuracy: 0.9142

17/30 [================>.............] - ETA: 0s - loss: 0.2191 - accuracy: 0.9135

19/30 [==================>...........] - ETA: 0s - loss: 0.2192 - accuracy: 0.9143

21/30 [====================>.........] - ETA: 0s - loss: 0.2199 - accuracy: 0.9141

23/30 [======================>.......] - ETA: 0s - loss: 0.2208 - accuracy: 0.9138

25/30 [========================>.....] - ETA: 0s - loss: 0.2206 - accuracy: 0.9141

27/30 [==========================>...] - ETA: 0s - loss: 0.2210 - accuracy: 0.9134

29/30 [============================>.] - ETA: 0s - loss: 0.2209 - accuracy: 0.9139

30/30 [==============================] - 1s 50ms/step - loss: 0.2204 - accuracy: 0.9141 - val_loss: 0.3103 - val_accuracy: 0.8649


Epoch 18/20


 1/30 [>.............................] - ETA: 0s - loss: 0.1966 - accuracy: 0.9355

 3/30 [==>...........................] - ETA: 0s - loss: 0.2038 - accuracy: 0.9264

 5/30 [====>.........................] - ETA: 0s - loss: 0.2025 - accuracy: 0.9250

 7/30 [======>.......................] - ETA: 0s - loss: 0.2035 - accuracy: 0.9213

 9/30 [========>.....................] - ETA: 0s - loss: 0.2069 - accuracy: 0.9206

11/30 [==========>...................] - ETA: 0s - loss: 0.2085 - accuracy: 0.9215

13/30 [============>.................] - ETA: 0s - loss: 0.2104 - accuracy: 0.9199

15/30 [==============>...............] - ETA: 0s - loss: 0.2121 - accuracy: 0.9185

17/30 [================>.............] - ETA: 0s - loss: 0.2106 - accuracy: 0.9184

19/30 [==================>...........] - ETA: 0s - loss: 0.2103 - accuracy: 0.9189

21/30 [====================>.........] - ETA: 0s - loss: 0.2090 - accuracy: 0.9187

23/30 [======================>.......] - ETA: 0s - loss: 0.2083 - accuracy: 0.9189

25/30 [========================>.....] - ETA: 0s - loss: 0.2079 - accuracy: 0.9197

27/30 [==========================>...] - ETA: 0s - loss: 0.2076 - accuracy: 0.9195

29/30 [============================>.] - ETA: 0s - loss: 0.2067 - accuracy: 0.9205

30/30 [==============================] - 1s 50ms/step - loss: 0.2066 - accuracy: 0.9206 - val_loss: 0.3061 - val_accuracy: 0.8655


Epoch 19/20


 1/30 [>.............................] - ETA: 0s - loss: 0.2163 - accuracy: 0.9219

 3/30 [==>...........................] - ETA: 0s - loss: 0.2064 - accuracy: 0.9238

 5/30 [====>.........................] - ETA: 0s - loss: 0.1992 - accuracy: 0.9254

 7/30 [======>.......................] - ETA: 0s - loss: 0.1993 - accuracy: 0.9286

 9/30 [========>.....................] - ETA: 0s - loss: 0.1967 - accuracy: 0.9282

11/30 [==========>...................] - ETA: 0s - loss: 0.1943 - accuracy: 0.9288

13/30 [============>.................] - ETA: 0s - loss: 0.1968 - accuracy: 0.9268

15/30 [==============>...............] - ETA: 0s - loss: 0.1988 - accuracy: 0.9262

17/30 [================>.............] - ETA: 0s - loss: 0.1971 - accuracy: 0.9273

19/30 [==================>...........] - ETA: 0s - loss: 0.1968 - accuracy: 0.9266

21/30 [====================>.........] - ETA: 0s - loss: 0.1942 - accuracy: 0.9279

23/30 [======================>.......] - ETA: 0s - loss: 0.1931 - accuracy: 0.9279

25/30 [========================>.....] - ETA: 0s - loss: 0.1932 - accuracy: 0.9268

27/30 [==========================>...] - ETA: 0s - loss: 0.1936 - accuracy: 0.9253

29/30 [============================>.] - ETA: 0s - loss: 0.1936 - accuracy: 0.9257

30/30 [==============================] - 1s 49ms/step - loss: 0.1941 - accuracy: 0.9256 - val_loss: 0.3050 - val_accuracy: 0.8706


Epoch 20/20


 1/30 [>.............................] - ETA: 0s - loss: 0.1883 - accuracy: 0.9277

 3/30 [==>...........................] - ETA: 0s - loss: 0.1971 - accuracy: 0.9290

 5/30 [====>.........................] - ETA: 0s - loss: 0.1903 - accuracy: 0.9324

 7/30 [======>.......................] - ETA: 0s - loss: 0.1875 - accuracy: 0.9286

 9/30 [========>.....................] - ETA: 0s - loss: 0.1888 - accuracy: 0.9264

11/30 [==========>...................] - ETA: 0s - loss: 0.1882 - accuracy: 0.9279

13/30 [============>.................] - ETA: 0s - loss: 0.1874 - accuracy: 0.9304

15/30 [==============>...............] - ETA: 0s - loss: 0.1894 - accuracy: 0.9299

17/30 [================>.............] - ETA: 0s - loss: 0.1898 - accuracy: 0.9300

19/30 [==================>...........] - ETA: 0s - loss: 0.1890 - accuracy: 0.9300

21/30 [====================>.........] - ETA: 0s - loss: 0.1868 - accuracy: 0.9306

23/30 [======================>.......] - ETA: 0s - loss: 0.1848 - accuracy: 0.9318

25/30 [========================>.....] - ETA: 0s - loss: 0.1850 - accuracy: 0.9320

27/30 [==========================>...] - ETA: 0s - loss: 0.1833 - accuracy: 0.9327

29/30 [============================>.] - ETA: 0s - loss: 0.1827 - accuracy: 0.9324

30/30 [==============================] - 2s 50ms/step - loss: 0.1824 - accuracy: 0.9323 - val_loss: 0.3021 - val_accuracy: 0.8714


## 评估模型

我们来看下模型的表现如何。将返回两个值。损失值（loss）（一个表示误差的数字，值越低越好）与准确率（accuracy）。

In [11]:
results = model.evaluate(test_data.batch(512), verbose=2)

for name, value in zip(model.metrics_names, results):
  print("%s: %.3f" % (name, value))

49/49 - 1s - loss: 0.3110 - accuracy: 0.8632


loss: 0.311
accuracy: 0.863


这种十分朴素的方法得到了约 87% 的准确率（accuracy）。若采用更好的方法，模型的准确率应当接近 95%。

## 进一步阅读

有关使用字符串输入的更一般方法，以及对训练期间准确率（accuracy）和损失值（loss）更详细的分析，请参阅[此处](https://tensorflow.google.cn/tutorials/keras/basic_text_classification)。